dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/EntityManagerBuilder.java

```diff
@@ -114,148 +119,175 @@ protected void collectEntityInfo(Set<Class<?>> entityTypes) throws Exception {
-                    attributeNamesForUpdate.remove(idAttrName);
-                if (versionAttrName != null)
-                    attributeNamesForUpdate.remove(versionAttrName);
 
-                SortedMap<String, Member> idClassAttributeAccessors = null;
-                if (!entityType.hasSingleIdAttribute()) {
-                    // Per JavaDoc, the above means there is an IdClass.
-                    // An EclipseLink extension that allows an Id on an embeddable of an entity
-                    // is an exception to this, which we indicate with idClassType null.
-                    Type<?> idClassType = getIdType(entityType);
-                    if (idClassType != null) {
-                        @SuppressWarnings("unchecked")
-                        Set<SingularAttribute<?, ?>> idClassAttributes = (Set<SingularAttribute<?, ?>>) (Set<?>) entityType.getIdClassAttributes();
-                        if (idClassAttributes != null) {
-                            attributeNames.remove(ID);
-                            idType = idClassType.getJavaType();
-                            idClassAttributeAccessors = getIdClassAccessors(idType, idClassAttributes);
+                    attributeNamesForUpdate.remove(ID);
+                    String idAttrName = attributeNames.get(ID);
+                    if (idAttrName != null)
+                        attributeNamesForUpdate.remove(idAttrName);
+                    if (versionAttrName != null)
+                        attributeNamesForUpdate.remove(versionAttrName);
+
+                    if (!entityType.hasSingleIdAttribute()) {
+                        // Per JavaDoc, the above means there is an IdClass.
+                        // An EclipseLink extension that allows an Id on an embeddable of an entity
+                        // is an exception to this, which we indicate with idClassType null.
+                        Type<?> idClassType = getIdType(entityType);
+                        if (idClassType != null) {
+                            @SuppressWarnings("unchecked")
+                            Set<SingularAttribute<?, ?>> idClassAttributes = (Set<SingularAttribute<?, ?>>) (Set<?>) entityType.getIdClassAttributes();
+                            if (idClassAttributes != null) {
+                                attributeNames.remove(ID);
+                                idType = idClassType.getJavaType();
+                                idClassAttributeAccessors = getIdClassAccessors(idType, idClassAttributes);
+                            }
                         }
                     }
+                } catch (MappingException e) {
+                    collectionException = e;
+                } catch (Exception e) {
+                    if (trace & tc.isDebugEnabled())
+                        Tr.debug(tc, "Unexpected exception caught while collecting entity information. The entity information will complete exceptionally", e);
+                    collectionException = e;
```

Comment:
This is a note for us to optimize this path and cover a related scenario later.  Don't worry about addressing it now.  It isn't related to your changes, but I happened to notice it while reviewing.
```suggestion
                    collectionException = e;
                    // TODO compute the userEntityClass earlier and completeExceptionally here.
                    // Then move the EntityInfo creation and successful completion to before the catch.
                    // That will also cover the possibility of EntityInfo constructor failing,
                    // which can happen when it doesn't validate.
```